### PR TITLE
Changes to compile hexagon runtimes on windows

### PR DESCRIPTION
**Pull request by ronlieb**

### CODE REVIEW

File: src/runtime/hexagon_host.cpp [hunk 1]
```diff
@@ -112,7 +112,13 @@ WEAK int init_hexagon_runtime(void *user_context) {
     // code execution based on the runtime; devices with Hexagon hardware will
     // simply provide conduits for execution on that hardware, while test/desktop/etc
     // environments can instead connect a simulator via the API.
-    void *host_lib = halide_load_library("libhalide_hexagon_host.so");
+    // If we are a windows hosted simulater, then we need to load the dll file.
```
Comment (dsharletg) [inline src/runtime/hexagon_host.cpp]: simulator

File: src/runtime/hexagon_host.cpp [hunk 2]
```diff
@@ -112,7 +112,13 @@ WEAK int init_hexagon_runtime(void *user_context) {
...
+    // Otherwise if we are a linux hosted simulator we need the .so.
+    void *host_lib = halide_load_library("libhalide_hexagon_host.dll");
```
Comment (dsharletg) [inline src/runtime/hexagon_host.cpp]: This really should try libhalide_hexagon_host.so first. That's the path that actually occurs in production on device, this is going to generate a guaranteed failed load library call every time on device.

File: src/runtime/HalideRuntime.h [hunk 3]
```diff
@@ -23,8 +23,10 @@ extern "C" {
 // it is not necessary, and may produce warnings for some build configurations.
 #ifdef _MSC_VER
 #define HALIDE_ALWAYS_INLINE __forceinline
+#define HALIDE_EXPORT __declspec(dllexport)
```
Comment (dsharletg) [inline src/runtime/HalideRuntime.h]: This should be added to sim_host.cpp, not HalideRuntime.h

File: src/runtime/hexagon_host.cpp [hunk 4]
```diff
@@ -112,7 +112,13 @@ WEAK int init_hexagon_runtime(void *user_context) {
     // code execution based on the runtime; devices with Hexagon hardware will
     // simply provide conduits for execution on that hardware, while test/desktop/etc
     // environments can instead connect a simulator via the API.
+    // If we are a windows hosted simulator, then we need to load the dll file.
+    // Otherwise if we are a linux hosted simulator we need the .so.
```
Comment (dsharletg) [inline src/runtime/hexagon_host.cpp]: It's really that if we aren't on the windows simulator, then we want the .so, whether it's linux or android (on device).

**Comment by ronlieb** [comment]
btw. i built locally the reordered library search 
1) open .so
2) if that returns NULL , try the .dll
This produces a windows dialogue window stating the dlopen of the .so failed, i close the dialogue and the program continues and runs correctly.

So, then i added a quick hack as follows to check for .so, and return NULL if present,
this prevents the LoadLibrary error dialogue box, and allows the test to pass.
There is probably a more elegant way to do this, bu the gist of it is, we need to NOT open the .so file if calling LoadLibrary.

WEAK void *halide_load_library_impl(const char *name) {
    if (strstr(name, ".so") != NULL) {
        return NULL;
    }
    return LoadLibraryA(name);
}

**Comment by ronlieb** [comment]
i google'ed around and found a more promising approach 'setErrorMode' 
trying that ...

**Comment by ronlieb** [comment]
I like this approach 

   unsigned oldMode = SetErrorMode(0);
    SetErrorMode(oldMode | SEM_FAILCRITICALERRORS | SEM_NOOPENFILEERRORBOX);
    void * Lib = LoadLibraryA(name);
    SetErrorMode(oldMode);

File: src/runtime/windows_get_symbol.cpp [hunk 5]
```diff
@@ -20,7 +23,12 @@ WEAK void *halide_get_symbol_impl(const char *name) {
 }
 
 WEAK void *halide_load_library_impl(const char *name) {
-    return LoadLibraryA(name);
+    // Suppress dialogoe windows during library open.
```
Comment (dsharletg) [inline src/runtime/windows_get_symbol.cpp]: dialog

File: src/runtime/windows_get_symbol.cpp [hunk 6]
```diff
@@ -20,7 +23,12 @@ WEAK void *halide_get_symbol_impl(const char *name) {
...
+    unsigned oldMode = SetErrorMode(0);
```
Comment (dsharletg) [inline src/runtime/windows_get_symbol.cpp]: Halide style is old_mode

File: src/runtime/windows_get_symbol.cpp [hunk 7]
```diff
@@ -20,7 +23,12 @@ WEAK void *halide_get_symbol_impl(const char *name) {
...
+    SetErrorMode(oldMode | SEM_FAILCRITICALERRORS | SEM_NOOPENFILEERRORBOX);
+    void * Lib = LoadLibraryA(name);
```
Comment (dsharletg) [inline src/runtime/windows_get_symbol.cpp]: Halide style is

    void *lib

Note removed space and lowercase lib.

**Comment by abadams** [comment]
Note that PRs are now being tested on all the buildbots too. This is new, so weird things may happen (e.g. unexplained failures because I'm cancelling builds or restarting the master or introducing bugs).

**Comment by ronlieb** [comment]
so did this sort of pass your buildbots? or is there an issue i need to look into?

**Comment by abadams** [comment]
It has passed on most of them, yeah. No errors. Still hasn't started the win-32 build, but the win-64 build was fine, so the chance of there being something wrong on win-32 is remote. We could wait for them all to finish, but we don't really have enough buildbot capacity to be picky about that yet. I'm acquiring more.

**Comment by ronlieb** [comment]
your call, it can wait a day or 2...

File: src/runtime/windows_get_symbol.cpp [hunk 8]
```diff
@@ -20,7 +23,12 @@ WEAK void *halide_get_symbol_impl(const char *name) {
 }
 
 WEAK void *halide_load_library_impl(const char *name) {
-    return LoadLibraryA(name);
+    // Suppress dialog windows during library open.
+    unsigned old_mode = SetErrorMode(0);
```
Comment (abadams) [inline src/runtime/windows_get_symbol.cpp]: Is this ever called in multi-threaded contexts? It doesn't seem thread-safe, as if errors are already suppressed, it leaves open a time in which errors could open a popup. Might be safer to do:

    unsigned old_mode = SetErrorMode(SEM_FAILCRITICALERRORS | SEM_NOOPENFILEERRORBOX);
    void *lib = LoadLibraryA(name);
    SetErrorMode(old_mode);
    return lib;

or grab a ScopedSpinLock on a global weak int.

**Pull request merged**